### DISCUSSION TITLE
How should we track maintanance tasks and feature implementatons?

### DISCUSSION THREAD

**Discussion by gasparnagy** [reqnroll/Reqnroll#26] [Contributor Q&A]
My original idea was that we keep the "Issues" list for public facing issues, that we should fix & close as quickly as possible, but for everyting else (feature implementartions, maintanance tasks, etc) we use the projects.

Now as I started the play with the new project system I realized that it has some massive UX issues (see [why is everything a draft?](https://github.com/orgs/community/discussions/10317)), so I'm not sure anymore. The suggested workaround is to make a separate repo, just for the tasks on the project board and convert every single "draft" card to an issue there.

I'm not sure... Is this really the way we want to go? Any thoughts / suggestions?

**Comment by ajeckmans** [comment]
I was wondering about this as well. Specifically to keep track on what needs to be done before a new release can be created. I think the work-around makes sense. Keeping the planning issues separate from real issues and questions.

Taking this a bit broader I think writing down some guidelines around merging into the main branch, steps to create a release, how to block a release, when to update dependencies would be welcome. Unless of course the current status quo of you doing all of the work is ok :)

**Reply by gasparnagy** [reply]
You are absolutely right, and we should deal with the processes and also document (and automate) them better.

I just always get back to the point that I first want to clear out testing and the target frameworks, because it influences the release (and any other) process heavily.

As a first step, I will try to setup a task board. At least we can also better track these meta-tasks, like cleaning up the release process better. We had to create a meta-project (https://github.com/reqnroll/.github) anyway, maybe that is the one we can use to host the task-issues.

**Reply by ajeckmans** [reply]
Hey, did you manage to set up that task board/project yet? It is kind of hard to keep track on what needs to be done for 2.0.0 to be released.

**Reply by gasparnagy** [reply]
No, not yet. I'm quite short in time this weeks. Do you have time maybe to help?

**Reply by ajeckmans** [reply]
I'll see what I can do this evening.

**Reply by ajeckmans** [reply]
I don't have the rights to set up new projects in the .github repo

**Reply by gasparnagy** [reply]
@ajeckmans changed the permissions. can you try now?

**Reply by ajeckmans** [reply]
yes, now I can. But now I wonder whether it wouldn't be a better idea to create the project at the organization level. As at least some of the maintenance and releases would impact/cover multiple repositories.

**Reply by gasparnagy** [reply]
Yes, this is what we will do, but technically this is not possible as the org level there is no "issue" storage that can hold the data for the tasks. But at the end it should work in a way that the project board will look like it is on org level, but internally it will store the data in the .github repo. I haven't tried that but only read about it that this is how the setup should look like.

**Reply by gasparnagy** [reply]
I don't know if it requires to have some special org-level permission to configure this.

**Reply by ajeckmans** [reply]
I don't think this is entirely true. You can track across repositories at the organization level. The only real issue is that with the current plan you can only auto-add from a single repository. Having said that, it is still better than nothing.
I've invited you to a WIP project to just track releases. As I can foresee the need to update documentation, sample projects, etc. having a project at the org level seems to be the most appropriate.

**Reply by gasparnagy** [reply]
> 've invited you to a WIP project to just track releases.

This is what you mean, right? (I haven't got a notification about that) https://github.com/orgs/reqnroll/projects/3 

> As I can foresee the need to update documentation, sample projects, etc. having a project at the org level seems to be the most appropriate.

Yes, agree. These are mostly org task.

**Reply by ajeckmans** [reply]
yes.

**Reply by ajeckmans** [reply]
I did not want to spam everyone until it is set up correctly, but core-team should probably be invited at some point so that everyone can see/edit it.

**Reply by gasparnagy** [reply]
So finally I think we need to set it up in a way that we can answer the following questions. Or with other words these are the task categories I can think of right now.

- How to add an org-level task, that is not related to a release (e.g. "Add sponsors page to website")
- How to add an org-level task, that is related to a release (e.g. "Extend the ReqOverflow sample with new DataTableHelpers after release of v2")
- How to add project-level refactoring tasks (e.g. "Change Specs project tests to run against in-memory implementation")
- How to add project-level feature tasks, that does not have a PR yet (e.g. "Support for property injection")
- How to add project-level issues or features with existing PR

**Reply by ajeckmans** [reply]
Yes, so my thoughts for this single view was to just track whatever is needed for the upcoming releases, where milestone equals release (quite obvious). And refactoring tasks are basically without milestone as well, but still put on the board and just assign them to a release when they're about done and we know which release it can go in.

Now a separate view might be focused around the non-release related tasks. We can either use labels to organize them, or we can add a custom field for that indicating what kind of work it is.

> How to add project-level feature tasks, that does not have a PR yet (e.g. "Support for property injection")
> How to add project-level issues or features with existing PR

If you add them as an issue, you can add them directly on the board, which is what I've done for the PRs. 
Otherwise I'd suggest adding a custom field `Discussion` and then add an item on the board, such as the current `Update xUnit Dependency...` one and set the Discussion field to the discussion url where it was discussed.

**Reply by ajeckmans** [reply]
Coming back to that. I think a custom field "Task Type" would be very helpful as well. If we were to group tasks into their type you can basically group/slice per task type. Which means in a view you can keep track of any non-release related long running tasks as well.

I added a second view that operates on the discussion field. If you administer that correctly, you'd basically be able to create a discussion as a meta item, and group all issues for that meta item.

**Reply by ajeckmans** [reply]
I think this is also one of those things you just need an initial view for and then see if it works or not.

**Reply by gasparnagy** [reply]
@ajeckmans looks good, but I need a fresh mind to process all what you said. Adding a task type might be something that is worth trying.

The "Update xunit dependency..." task is now a "draft" task. I've read somewhere, that keeping the tasks that way is not good, because you cannot have tags on them and filter for them(?). Do you know about that problem?

**Reply by gasparnagy** [reply]
Draft tasks: https://github.com/orgs/community/discussions/10317

**Reply by ajeckmans** [reply]
That is correct, but you can add custom project fields to them. Which is what the Discussion column now points to. And even though you cannot filter on them, you can group, slice and filter on them in the projects view and they also do not surface in the original issues, PRs or discussions as they basically do not 'exist' (they do on issues, but only in the column on the right).
I've added a view of how you could use them to group tasks like that: https://github.com/orgs/reqnroll/projects/3/views/5

**Reply by ajeckmans** [reply]
This is my first time working with Projects in Github, and I definitely see some shortcomings here and there. However there is some flexibility where you can get to some kind of working process/overview without adapting too much to Githubs preferred way of doing things.

**Reply by gasparnagy** [reply]
First time for me as well. I had a quick look and it is fine. A few smaller questions/problems I have noticed:

- Can we set default values for the fields? It is now confusing what is the difference between the not set status and the "ToDo".
- I wonder if we should just filter out the "Done" tasks from all views. We anyway have the changelog as the ultimate log of what happened. So the done tasks in this list might only be disturbing.
- I created a task `Simplify target platform of Reqnroll to use .NET Standard 2.0 only`, but it seems that in this form I cannot assign it to the v2 milestone.

**Reply by ajeckmans** [reply]
That is correct. If you create a draft issue you cannot assign it to a milestone. However especially for those kinds of tasks I think that is the correct behavior. Either it has no milestone attached, which means it won't block a release of that milestone, or it has one and it will need a PR or issue anyways.

I agree with filtering all of the done tasks. 

Default value is set through the workflows (https://github.com/orgs/reqnroll/projects/3/workflows). This applies to all newly added items, but not the existing ones, but I've gone ahead and set it manually.

**Reply by gasparnagy** [reply]
My original idea was that we don't use the issues (of the Reqnroll repo) for features, because it pollutes the issue list (an open "issue" for longer time distructs our focus on the real issues that we want to fix quickly), but only use discussion items for these until we start working on them and they will have a PR. It can happen however, that we know that we want them to be included in a milestone although the implementation has not started yet. (My example of task `Simplify target platform of Reqnroll to use .NET Standard 2.0 only` is exactly like that.)

Arranging these to a milestone is maybe not that critical, so we can stay with how it is now, or we can also try to experiment with creating a task-issue for them in the `.github` repo. The third option would be to make a pseudo PR for these (a PR with an empty commit).

**Reply by ajeckmans** [reply]
That or we abandon the milestone completely and add a custom field 'target release version' to keep track of it all.

**Reply by gasparnagy** [reply]
Oh, that's a good idea. Let's do that for now at least!

**Reply by clrudolphi** [reply]
To clarify on how to track bug fixes...
Presume someone identifies a problem and enters a discussion item about it. We then create an Issue to track it, and eventually a PR to fix it.
Should all three entries be shown on the project tracker?
Which of those would be assignable to a release milestone?
After the PR is approved and merged, when should any of these be marked as completed - at the time of PR merge or when the Release is published?

**Reply by gasparnagy** [reply]
@clrudolphi These are all good questions and I don't know yet the answer - we need to see how it works. But in general, if someone identifies an issue and we simply fix it, we don't need to assign it to milestones. I would do the milestone assignment only for long-running, long-pending issues (which we should generally avoid). For the smaller ones we have the changelog anyway that is a precise description of what we have done.
